pkg/generate/app/pipeline.go

```diff
@@ -180,43 +181,62 @@ func (s sortablePorts) Swap(i, j int) {
 	s[j] = p
 }
 
+// AddServices sets up services for the provided objects
 func AddServices(objects Objects) Objects {
 	svcs := []runtime.Object{}
 	for _, o := range objects {
 		switch t := o.(type) {
 		case *deploy.DeploymentConfig:
 			for _, container := range t.Template.ControllerTemplate.Template.Spec.Containers {
+				name, generateName := makeValidServiceName(t.Name)
+				svc := &kapi.Service{
+					ObjectMeta: kapi.ObjectMeta{
+						Name:         name,
+						GenerateName: generateName,
+						Labels:       t.Labels,
+					},
+					Spec: kapi.ServiceSpec{
+						Selector: t.Template.ControllerTemplate.Selector,
+					},
+				}
+
 				ports := sortablePorts(container.Ports)
 				sort.Sort(&ports)
 				for _, p := range ports {
-					name, generateName := makeValidServiceName(t.Name)
-					svcs = append(svcs, &kapi.Service{
-						ObjectMeta: kapi.ObjectMeta{
-							Name:         name,
-							GenerateName: generateName,
-							Labels:       t.Labels,
-						},
-						Spec: kapi.ServiceSpec{
-							Selector: t.Template.ControllerTemplate.Selector,
-							Ports: []kapi.ServicePort{
-								{
-									Name:       p.Name,
-									Port:       p.ContainerPort,
-									Protocol:   p.Protocol,
-									TargetPort: kutil.NewIntOrStringFromInt(p.ContainerPort),
-								},
-							},
-						},
+					svc.Spec.Ports = append(svc.Spec.Ports, kapi.ServicePort{
+						Name:       p.Name,
+						Port:       p.ContainerPort,
+						Protocol:   p.Protocol,
+						TargetPort: kutil.NewIntOrStringFromInt(p.ContainerPort),
 					})
-					break
 				}
-				break
+				svcs = append(svcs, svc)
 			}
 		}
 	}
 	return append(svcs, objects...)
 }
 
+// AddRoutes sets up routes for the provided objects
+func AddRoutes(objects Objects) Objects {
+	routes := []runtime.Object{}
+	for _, o := range objects {
+		switch t := o.(type) {
+		case *kapi.Service:
+			for range t.Spec.Ports {
+				routes = append(routes, &route.Route{
+					ObjectMeta: kapi.ObjectMeta{
+						Name:   t.Name,
```

Comment:
I suspect routes should declare named bindings for protocols:

"http": 8080,  # container port
"https": 8443, # container port

and the router decides what to do with "http" (it either supports it or not) and then the router scheduler will update the status of the route to say what actual port and IP and DNS address that port gets exposed on.

I.e.

Pod exposes 8080, 8443
Service exposes 8443
Route points to service, defines "http":8080, "https":8443
Route Allocator binds Route to Router Foo
Router Foo listens on 80 for "http" and 443 for "https", so the status of the route gets set to:

"status": {
  "router": "foo",
  "bindings": [{
    "name": "http",
    "port": "80",
    "dns": "route-namespace.myrouter.west.com"
  }]
}

None of that is there today, needs more thought and design.

----- Original Message -----
> Routes will have to have a port mapping, but routers decide what outbound
> ports you get (@pweil-).  I.e., a user can say "I want to expose port 443
> from containers via the router", but only the router knows what port that
> actually is.
> 
> ----- Original Message -----
> > >  			}
> > >  		}
> > >  	}
> > >  	return append(svcs, objects...)
> > >  }
> > >  
> > > +// AddRoutes sets up routes for the provided objects
> > > +func AddRoutes(objects Objects) Objects {
> > > +	routes := []runtime.Object{}
> > > +	for _, o := range objects {
> > > +		switch t := o.(type) {
> > > +		case *kapi.Service:
> > > +			for range t.Spec.Ports {
> > > +				routes = append(routes, &route.Route{
> > > +					ObjectMeta: kapi.ObjectMeta{
> > > +						Name:   t.Name,
> > 
> > @smarterclayton here as well, every route gets its name from the service it
> > will connect to. But what happens with a multiport service? Every route on
> > it will have the same name as is.
> > 
> > ---
> > Reply to this email directly or view it on GitHub:
> > https://github.com/openshift/origin/pull/1747/files#r28770122